include/deal.II/matrix_free/evaluation_selector.h

```diff
@@ -0,0 +1,574 @@
+
+  /**
+   * This class chooses the correct template degree.
+   */
+  template<int degree, int n_q_points_1d, int dim, int n_components, typename Number>
+  struct Factory<dim, n_components, Number, 0, degree, n_q_points_1d>
+  {
+    static inline void evaluate (
+      const internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<Number> > &shape_info,
+      VectorizedArray<Number> *values_dofs_actual[],
+      VectorizedArray<Number> *values_quad[],
+      VectorizedArray<Number> *gradients_quad[][dim],
+      VectorizedArray<Number> *hessians_quad[][(dim*(dim+1))/2],
+      VectorizedArray<Number> *scratch_data,
+      const bool               evaluate_values,
+      const bool               evaluate_gradients,
+      const bool               evaluate_hessians)
+    {
+      const unsigned int runtime_degree = shape_info.fe_degree;
+      constexpr unsigned int nonnegative_start_n_q_points = (degree>0)?degree-1:0;
+      if (runtime_degree == degree)
+        Factory<dim, n_components, Number, 1, degree, nonnegative_start_n_q_points>::evaluate
+        (shape_info, values_dofs_actual, values_quad, gradients_quad, hessians_quad,
+         scratch_data, evaluate_values, evaluate_gradients, evaluate_hessians);
+      else
+        Factory<dim, n_components, Number, 0, degree+1, n_q_points_1d>::evaluate
+        (shape_info, values_dofs_actual, values_quad, gradients_quad, hessians_quad,
+         scratch_data, evaluate_values, evaluate_gradients, evaluate_hessians);
+    }
+
+    static inline void integrate (
+      const internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<Number> > &shape_info,
+      VectorizedArray<Number> *values_dofs_actual[],
+      VectorizedArray<Number> *values_quad[],
+      VectorizedArray<Number> *gradients_quad[][dim],
+      VectorizedArray<Number> *scratch_data,
+      const bool               integrate_values,
+      const bool               integrate_gradients)
+    {
+      const int runtime_degree = shape_info.fe_degree;
+      constexpr unsigned int nonnegative_start_n_q_points = (degree>0)?degree-1:0;
+      if (runtime_degree == degree)
+        Factory<dim, n_components, Number, 1, degree, nonnegative_start_n_q_points>::integrate
+        (shape_info, values_dofs_actual, values_quad, gradients_quad,
+         scratch_data, integrate_values, integrate_gradients);
+      else
+        Factory<dim, n_components, Number, 0, degree+1, n_q_points_1d>::integrate
+        (shape_info, values_dofs_actual, values_quad, gradients_quad,
+         scratch_data, integrate_values, integrate_gradients);
+    }
+  };
+
+  /**
+   * This class chooses the correct template n_q_points_1d after degree was chosen.
+   */
+  template<int degree, int n_q_points_1d, int dim, int n_components, typename Number>
+  struct Factory<dim, n_components, Number, 1, degree, n_q_points_1d,
+    typename std::enable_if<n_q_points_1d<degree+2>::type>
```

Comment:
I find this clause hard to read. Can you wrap `n_q_points_1d<degree+2` into `()` to highlight that we look at a boolean condition and avoid ambiguity regarding the template delimiters?